src/coreclr/tools/aot/ILCompiler.RyuJit/JitInterface/CorInfoImpl.RyuJit.cs

```diff
@@ -2203,5 +2203,74 @@ private int getExactClasses(CORINFO_CLASS_STRUCT_* baseType, int maxExactClasses
             Debug.Assert(index <= maxExactClasses);
             return index;
         }
+
+        private bool getReadonlyStaticFieldValue(CORINFO_FIELD_STRUCT_* fieldHandle, byte* buffer, int bufferSize)
+        {
+            FieldDesc field = HandleToObject(fieldHandle);
+            if (field.IsStatic && !field.IsThreadStatic && field.IsInitOnly && field.OwningType is MetadataType owningType)
+            {
+                PreinitializationManager preinitManager = _compilation.NodeFactory.PreinitializationManager;
+                if (preinitManager.IsPreinitialized(owningType))
+                {
+                    TypePreinit.ISerializableValue value = preinitManager
+                        .GetPreinitializationInfo(owningType).GetFieldValue(field);
+
+                    if (value == null)
+                    {
+                        *((nint*)buffer) = 0;
+                        return true;
+                    }
+
+                    if (value.GetRawData(_compilation.NodeFactory, out object data))
+                    {
+                        switch (data)
+                        {
+                            case byte[] bytes when bytes.Length <= bufferSize:
+                                bytes.AsSpan().CopyTo(new Span<byte>(buffer, bufferSize));
+                                return true;
+
+                            case FrozenObjectNode or FrozenStringNode:
+                                *(nint*)buffer = ObjectToHandle(data);
+                                return true;
+                        }
+                    }
+                }
+            }
+            return false;
+        }
+
+        private CORINFO_CLASS_STRUCT_* getObjectType(void* objPtr)
+        {
+            object obj = HandleToObject((IntPtr)objPtr);
+            return obj switch
+            {
+                // For now we only support frozen strings
+                FrozenStringNode => ObjectToHandle(_compilation.TypeSystemContext.GetWellKnownType(WellKnownType.String)),
+
+                // and frozen objects like arrays
+                FrozenObjectNode frozenObj => ObjectToHandle(frozenObj.ObjectType),
+
+                _ => throw new NotImplementedException($"Unexpected object in getObjectType: {obj}")
+            };
+        }
+
+#pragma warning disable CA1822 // Mark members as static
+        private void* getRuntimeTypePointer(CORINFO_CLASS_STRUCT_* cls)
+#pragma warning restore CA1822 // Mark members as static
+        {
+            // TODO: https://github.com/dotnet/runtime/pull/75573#issuecomment-1250824543
+            return null;
+        }
+
+        private bool isObjectImmutable(void* objPtr)
+        {
+            object obj = HandleToObject((IntPtr)objPtr);
+            return obj switch
+            {
+                FrozenStringNode => true,
+                FrozenObjectNode => false,
```

Comment:
Empty arrays and object with no instance fields should be reported as Immutable.

@VSadov Is it ok to treat delegates as immutable? In other words, does "release with respect to write operations to the instance’s fields and metadata." apply transitively to what's referenced by the instance's fields or not?

Reply:
Empty array should work the same as string. If we create/fetch under a lock, we can pass it around with ordinary assignments.

Reply:
Same for static delegates that have not instance attached. 

If there is an instance, it can only be a readonly reference to another immutable object, because GC does not always trace into children of frozen objects. 
Then the same applies - if publishing happen in native code with appropriate fences (a lock would do that), then we can just pass a frozen delegate around as fully published immutable thing.

Reply:
> if publishing/sharing happens in native code under a lock

Yes, the frozen objects are always created under a lock.

> If there is an instance, it can only be a readonly reference to another immortal object

The immortal can have mutable state. Consider these two examples:

```csharp
sealed class InstanceWithoutField
{
    static readonly static Action a = new Action(new InstanceWithoutField().InstanceMethod);
    static static Action b;
    
    void InstanceMethod()
    {
       ...
    }
    
    static void Method()
    {
        b = a; // I assume that this does not need to be `stlr`. Correct?
    }
}

sealed class InstanceWithField
{
    public int X;

    static readonly static Action a = new Action(new InstanceWithField().InstanceMethod);
    static static Action b;
    
    void InstanceMethod()
    {
       ...
    }
    
    static void Method()
    {
        b = a; // I am not sure whether this needs to be `stlr`. Could you please clarify?
    }
}
```

Reply:
```cs
 // Assume that this instance is created as frozen object
 static readonly static Action a = new Action(new InstanceWithField().InstanceMethod);
```
If user can re-bind the delegate to a different instance we are in a bigger trouble as the new instance may get collected. I do not remember if it is possible to re-bind a delegate after creation.

If the instance is readonly, no need for stlr. If the whole thing created/shared under a lock, no fields will be missing when another thread accesses the delegate.

Reply:
It is not possible to re-bind delegate to a different instance.

Yes, no delegate fields will be missing when another thread accesses the delegate. But what about the fields of the instance that the delegate points to? Here is one more example:
```
sealed class InstanceWithField
{
    public int X;

    // Assume that this instance is created as frozen object;
    static readonly InstanceWithField inst = new InstanceWithField();
    static readonly Action a = new Action(inst.InstanceMethod);

    static Action b = new Action(new InstanceWithField() { X = 42 });
    
    void InstanceMethod()
    {
       Console.WriteLine(X); // I assume that this can print 0. Correct? 
    }
    
    static void MethodCalledOnThread1()
    {
        inst.X = 1; // I assume that this update is not guaranteed to be seen before `b = a` assignment is seen by other thread. Correct?
        b = a;
    }

    static void MethodCalledOnThread2()
    {
        b();
    }
}
```

Reply:
(An example like this may be worth mentioning in the memory model doc.)

Reply:
Ah, yes. The instance could be completely readonly, but if it refers to another frozen object, that must be readonly too. Or the whole thing is not readonly. - In a sense that passing it to another thread needs stlr or the other thread may see inconsistent field values.

The immutability must be recursive. Or maybe just treat objects with reference fields as not immutable.

Reply:
Ok, so this can return true for empty arrays, object with no instance fields and static delegates.

Reply:
Strings too?

Reply:
Yes, strings too.

Reply:
> Ok, so this can return true for empty arrays, object with no instance fields and static delegates.

On the VM side this returns true for `RuntimeType` as well - do we consider delegates and RuntimeType special cases, or are there more broadly applicable rules that delegates and RuntimeType fall into? (e.g. types that don't have mutable GC references, types that only have readonly fields?)

Reply:
Static delegates and RuntimeTypes are special cases since we have full control over the implementation and do not depend on the store implying release.

> types that only have readonly fields?

readonly on instance fields does not provide interesting guarantees for the runtime.

Reply:
Thanks, marked empty arrays, objects without fields and delegates as immutable via https://github.com/dotnet/runtime/pull/76112/commits/afefe442872705e6a413c391bec28541d82b8928